### PR TITLE
PERF: Improve replace perf

### PR DESCRIPTION
**Pull request by sinhrks**
- [x] tests added / passed
- [x] passes `git diff upstream/master | flake8 --diff`
- [x] whatsnew entry

When `.replace` is called with `dict`, replacements are done per value. Current impl try to soft convert the dtype in every replacement, but it is enough to be done in the final replacement.
#### Bench

```
-  712.83ms   355.93ms      0.50  replace.replace_convert.time_replace_series_timestamp
-     1.50s   698.21ms      0.46  replace.replace_convert.time_replace_frame_timestamp
-     3.12s   690.48ms      0.22  replace.replace_convert.time_replace_frame_timedelta
-     1.69s   354.83ms      0.21  replace.replace_convert.time_replace_series_timedelta
```

### CODE REVIEW

**Comment by jreback** [comment]
notice that ``BoolBlock.replace`` doesn't have ``convert`` arg at all! so not tested / working. can you give a check.

**Comment by sinhrks** [comment]
Ah, found some ``.replace`` conversion looks buggy. I'll submit a separate issue as the fix not looks straightforward.

File: pandas/tests/series/test_replace.py [hunk 1]
```diff
@@ -0,0 +1,276 @@
+# coding=utf-8
+# pylint: disable-msg=E1101,W0612
```
Comment (jreback) [inline pandas/tests/series/test_replace.py]: great!

**Comment by sinhrks** [comment]
See #12747. Change this PR title to WIP until #12747 is fixed.

**Comment by jreback** [comment]
can you rebase / update?

**Comment by jreback** [comment]
lgtm. can we merge this independent of the other PR's? (so it could go in 0.19.2)

**Comment by sinhrks** [comment]
It is independent. #12747 fixes some existing bugs separately.

**Comment by jreback** [comment]
thanks!

**Pull request closed without merging**